### ISSUE TITLE
About "pubfig65_patch_neuron_distance.py"

### ISSUE THREAD

**Issue by chnyangs** [bolunwang/translearn#3]
Hi Bolun:
Thanks for publishing the code of your paper. We are trying to reproduce your results.
Currently, we try to run "pubfig65_patch_neuron_distance.py" with the configurations as indicated in "README.md". However, there is an error when the testing data is sent to the model for prediction. It also does not work with your provided test file "pubfig65_imagenet_test.h5".
The screenshot is attached for your reference. We would be extremely grateful if you could take a look at this issue.

![Screenshot](https://user-images.githubusercontent.com/17538930/56448856-58395380-6356-11e9-9aa9-39c61330c98e.png)

Best,
xiangwen

**Comment by bolunwang** [comment]
Is there anything that could help repro this error?

**Comment by chnyangs** [comment]
Hi Bolun:
This issue happens when executing "eval_model" method which I using the data "pubfig65_imagenet_test.h5".
                 # evaluate model performance before training
Line 347    eval_model(student_model, X_test, Y_test, X_adv, Y_target)

Best,
Xiangwen

**Comment by chnyangs** [comment]
> Hi Bolun:
> This issue happens when executing "eval_model" method which I using the data "pubfig65_imagenet_test.h5".
> # evaluate model performance before training
> Line 347 eval_model(student_model, X_test, Y_test, X_adv, Y_target)
> 
> Best,
> Xiangwen

**Comment by chnyangs** [comment]
Hi Boun:
One more thing, the STUDENT_MODEL_FILE = 'models/pubfig65-vggface-trans-nb-train-90.h5' in "pubfig65_fingerprint_vggface.py"and 
STUDENT_MODEL_FILE = 'models/pubfig65_vggface_trans_nbtrain_90.h5' in "pubfig65_patch_neuron_distance.py" is the same model, right? Because I found a little difference between the model name.

By the way, Could you share me with the latest code and all the related datasets? That's will be a great help for my study.
Best,
Xiangwen

**Comment by bolunwang** [comment]
Well... that doesn't really help repro the error... Could you give a bit more context of what you executed? Could you paste the code snippet and double check the size of the input data? Based on the error message, it's probably the input data dimension is different from what the model expects. I'm not sure if this is because the Keras API has changed or you've modified the input data. But based on what you provided, I can't tell the real cause of this.

As for code and dataset, this repo already has all the key implementation of this project, and all of the datasets we use are public. You could find all the info you need in the readme.

**Comment by Bang505** [comment]
Hi Bolun:
I am also confusing about this code. And I think the function "transform_dataset()" which transforms the datasets causes the issue. But I do not know how to fix it.
In "pubfig65_patch_neuron_distance.py" Line 342-347:

>     X_train, Y_train = transform_dataset(bottleneck_model, X_train, Y_train, 0)
>     X_test, Y_test = transform_dataset(bottleneck_model, X_test, Y_test, 0)
>     X_adv, Y_target = transform_dataset(bottleneck_model, X_adv, Y_target, 0)
> 
>     # evaluate model performance before training
>     eval_model(student_model, X_test, Y_test, X_adv, Y_target)

X_train and X_test are all transformed by function "transform_dataset". This function adds a second part to X_test:
In "pubfig65_patch_neuron_distance.py" Line 287-294:

>     Y_pred = bottleneck_model.predict(X)
>     X_bottleneck = Y_pred
>     X = [X, X_bottleneck]
>     Y = [Y, np.array([cap] * Y.shape[0])]
>     return X, Y

After being transformed, these datasets are used in the function "eval_model". And in the function, X_test is directly feed to the student model while the second part does not fit the model input. I think this may cause the problem and it is what I am confusing. 
In "pubfig65_patch_neuron_distance.py" Line 305-313:

> def eval_model(student_model, X_test, Y_test, X_adv, Y_adv):
> 
>     # evaluate model classification accuracy and attack success rate
>     Y_pred = student_model.predict(X_test)
>     correct_indices = (
>         np.argmax(Y_pred[0], axis=1) ==
>         np.argmax(Y_test[0], axis=1))
>     test_acc = np.mean(correct_indices)
>     test_neuron_sim = np.mean(Y_pred[1])

Could you mind telling me whether I am misunderstanding your code? If yes,  could you mind telling me how this code works and how to fix this problem?

Best Regards,

**Comment by bolunwang** [comment]
Could you be a bit more specific about what the "issue" is?

I can help explain the code if that helps.

The idea of this defense is to add a second term into the loss function, so we need to modify the model, input format (to add the original bottleneck neuron vector), and the output (to add the cap). In the last code snippet, we just need to evaluate model's accuracy, so only the raw input (first part of X), and the label (first part of Y) is needed.

**Comment by Bang505** [comment]
Thanks for your kind help.
The issue is the same as xiangwen, here comes an error when the it goes to Line 347:

> eval_model(student_model, X_test, Y_test, X_adv, Y_target)

For this function, it will feed "X_test" into the "student_model" and get "Y_pred" where "X_test = [X, X_bottleneck]" after being transformed.

> Y_pred = student_model.predict(X_test)

As a result, X is the input image while X_bottleneck is the internal representation. Both of them are feed into student model. The output for X is the prediction of input images. What should be the output for X_bottleneck and how should we build one student model corresponding to these two different inputs?
Thanks for your time.

**Comment by Bang505** [comment]
Solved the problem. I failed to correctly build the data based on the code and the version of tensorflow. Thanks!
